src/main/java/jenkins/plugins/git/GitPublisherStep.java

```diff
@@ -0,0 +1,367 @@
+    this.branchesToPush = branchesToPush;
+  }
+
+  @DataBoundSetter
+  public void setNotesToPush(List<GitPublisher.NoteToPush> notesToPush) {
+    this.notesToPush = notesToPush;
+  }
+
+  public SCM createSCM() {
+    return new GitSCM(GitSCM.createRepoList(url, credentialsId), Collections.singletonList(new BranchSpec("*/" + branch)), false, Collections.<SubmoduleConfig>emptyList(), null,
+      null, Collections.<GitSCMExtension>singletonList(new LocalBranch(branch)));
+  }
+
+  public boolean isForcePush() {
+    return forcePush;
+  }
+
+  public boolean isPushTags() {
+    if (tagsToPush == null) {
+      return false;
+    }
+    return !tagsToPush.isEmpty();
+  }
+
+  public boolean isPushBranches() {
+    if (branchesToPush == null) {
+      return false;
+    }
+    return !branchesToPush.isEmpty();
+  }
+
+  public boolean isPushNotes() {
+    if (notesToPush == null) {
+      return false;
+    }
+    return !notesToPush.isEmpty();
+  }
+
+  public List<GitPublisher.TagToPush> getTagsToPush() {
+    if (tagsToPush == null) {
+      tagsToPush = new ArrayList<GitPublisher.TagToPush>();
+    }
+
+    return tagsToPush;
+  }
+
+  public List<GitPublisher.BranchToPush> getBranchesToPush() {
+    if (branchesToPush == null) {
+      branchesToPush = new ArrayList<GitPublisher.BranchToPush>();
+    }
+
+    return branchesToPush;
+  }
+
+  public List<GitPublisher.NoteToPush> getNotesToPush() {
+    if (notesToPush == null) {
+      notesToPush = new ArrayList<GitPublisher.NoteToPush>();
+    }
+
+    return notesToPush;
+  }
+
+
+  public static final class StepExecutionImpl extends AbstractSynchronousNonBlockingStepExecution<Void> {
+
+    @javax.inject.Inject
+    private transient GitPublisherStep step;
+    @StepContextParameter
+    private transient Run<?, ?> run;
+    @StepContextParameter
+    private transient FilePath workspace;
+    @StepContextParameter
+    private transient TaskListener listener;
+    @StepContextParameter
+    private transient Launcher launcher;
+
+    @Override
+    protected Void run() throws Exception {
+      step.publish(run, workspace, listener, launcher);
+      return null;
+    }
+
+    private static final long serialVersionUID = 1L;
+  }
+
+  public final void publish(Run<?, ?> run, FilePath workspace, TaskListener listener, Launcher launcher) throws Exception {
+    final GitSCM gitSCM = (GitSCM) createSCM();
+
+    EnvVars environment = run.getEnvironment(listener);
+
+    final GitClient git = gitSCM.createClient(listener, environment, run, workspace);
+
+    URIish remoteURI;
+
+    if (isPushTags()) {
+      for (final GitPublisher.TagToPush t : tagsToPush) {
+        if (t.getTagName() == null)
+          throw new AbortException("No tag to push defined");
+
+        if (t.getTargetRepoName() == null)
+          throw new AbortException("No target repo to push to defined");
+
+        final String tagName = environment.expand(t.getTagName());
```

Comment:
Formally the checks above do not prevent the return value from being `null` here. E.g. in the case of the parallel config change. I would recommend assigning the method's return value to a local variable before the check and tthen using this variable in the code